api/filter/accesslog.proto

```diff
@@ -53,120 +82,132 @@ message AccessLogEntry {
+  // route created from a higher level forwarding rule with some ID can place
+  // that ID in this field and cross reference later. It can also be used to
+  // determine if a canary endpoint was used or not.
+  google.protobuf.Struct metadata = 13;
+}
 
-  // The Upstream Cluster that the upstream host belongs to.
-  string upstream_cluster = 6;
+message TLSProperties {
+  // TLS Version
+  enum TLSVersion {
+    VERSION_UNSPECIFIED = 0;
+    TLSv1 = 1;
+    TLSv1_1 = 2;
+    TLSv1_2 = 3;
+    TLSv1_3 = 4;
+  }
+  TLSVersion tls_version = 1;
 
-  // This field is the IP and port on which the request from the user was
-  // received, stored in ipv4:port or [ipv6]:port format.
-  string destination_host = 7;
+  // TLS Cipher suite negotiated during TLS handshake.
+  // The value is four hex digits defined by the IANA TLS Cipher Suite Registry,
+  // eg, "009C" for TLS_RSA_WITH_AES_128_GCM_SHA256.
+  //
+  // Here it is expressed as an integer.
+  google.protobuf.UInt32Value tls_cipher_suite = 2;
 
-  // Size of the HTTP request body in bytes
-  google.protobuf.UInt64Value request_body_bytes = 8;
+  // SNI hostname from handshake.
+  string tls_sni_hostname = 3;
+}
 
-  // Size of the HTTP response body in bytes
-  google.protobuf.UInt64Value response_body_bytes = 9;
+message TCPAccessLogEntry {
+  // Common properties shared by all Envoy access logs.
+  AccessLogBaseEntry base_entry = 1;
+}
 
-  // Size of the HTTP request headers in bytes
-  google.protobuf.UInt64Value request_headers_bytes = 10;
+message HTTPRequestProperties {
+  // The request method (RFC 7231/2616)
+  RequestMethod request_method = 1;
 
-  // Size of the HTTP response headers in bytes
-  google.protobuf.UInt64Value response_headers_bytes = 11;
+  // The scheme portion of the incoming request URI
+  string scheme = 2;
 
-  // Whether the request arrived via a secure (TLS) protocol
-  google.protobuf.BoolValue secure = 12;
+  // HTTP2 :authority header value or HTTP1.1 Host header value
+  string authority = 3;
 
-  // Whether the request is a HealthCheck request
-  google.protobuf.BoolValue health_check = 13;
+  // The port of the incoming request URI
+  // (Unused currently, as port is composed onto authority)
+  google.protobuf.UInt32Value port = 4;
 
-  // The HTTP response code
-  google.protobuf.UInt32Value response_code = 14;
+  // The path portion from the incoming request URI
+  string path = 5;
 
-  // User agent as sent by client HTTP
-  string user_agent = 15;
+  // HTTP request headers
+  // ====================
 
-  // Path
-  //
-  // This is the Path portion from the incoming request URI
-  string path = 17;
+  // User agent as sent by client HTTP
+  string user_agent = 6;
 
   // Referer header as sent by client HTTP
   // (Referer is spelled to match the HTTP spec, not English).
-  string referer = 18;
+  string referer = 7;
 
   // X-Forwarded-For request header
-  string forwarded_for = 19;
+  string forwarded_for = 8;
```

Comment:
I know some headers are more "interesting" than others, but should we keep it simpler and put all of the headers into "repeated HeaderValue request_headers"?  (This, request_id, original_path).

Reply:
I don't have strong feelings here. The current approach strikes me as slightly more obvious/discoverable to someone who's unfamiliar with what the logs contain, but the tradeoff is the schema is a little noisier. I'm fine with either approach.

Reply:
My weak preference is to be explicit. Then the user can only pull the values they care about w/o iterating.

Reply:
I feel like what we have here is kind of awkward.  If what you want to log is one of the top-level http headers (like this one), it's fine, and you can log quickly/easily.  If you want to log some other header, you have to iterate through the header list.

It seems like we should do one of two things:
1) Put the rest of the headers in a map (https://developers.google.com/protocol-buffers/docs/proto3#maps) so they can be looked up easily

2) Have some way to specify which headers should be included in each log message; this prevents the log messages from getting bloated by large headers that we don't want to log.

If we do either of the above, it might make sense to treat these "special" headers like the rest and make the un-special.  Or maybe not.  But the argument that we need these here to avoid iterating makes me feel like the interface is wrong and/or too specialized.

Thoughts?

Reply:
We definitely need config on what special headers to log. I don't think we were going to log all headers, so that is not yet defined/implemented. So yes I agree we definitely need (2).

For (1) I'm fine moving to a map. I don't have that strong of an opinion on whether to then put certain headers always in the map. We could say that Envoy will guarantee to put some number of headers into the map and then any others will be specified by a future configuration mechanism. @goaway what do you think?

Reply:
What about metadata? This has a similar tradeoff around log verbosity and configuration.

Reply:
The difference there to me is that metadata is truly opaque. We do know we need certain headers in all cases. I don't really feel that strongly about this so am happy to go either way.

Reply:
I think the best approach would be to change the repeated header fields in this proto into a map used for custom headers not already covered by the named fields in the proto. The named fields (like "authority") would always be populated if relevant to the request.

All the custom headers desired to be put into the map must be explicitly configured. There could be a request/response pair of repeated string fields in the AccessLogServiceConfig proto which specify the desired headers to put into the custom headers map.

Reply:
Sounds like we have a consensus. I'm good with this approach (map for headers specified via configuration, alongside named fields for "high-significance" headers we always capture). Will update.

Reply:
sgtm

Reply:
+1

Reply:
I made these `map<string, string>`. I could have used `google.protobuf.Struct`, but it raises a bunch of questions about what it means for HTTP headers to have types and whether types should be invariant for a given header name/key, whether they should be specified in configuration or interpreted dynamically, what happens if we see an unexpected type, etc. And also this is a log format. But if someone feels strongly, I can change it.